andela/ah-backend-sparta
# #164798203 Enable user signup using social authentication

#### What does this PR do?
- Enable user to login via social authentication

#### Description of Task to be completed?
- The user should able to login using social media services like facebook, google and twitter

#### How should this be manually tested?
- Fetch branch `git fetch origin ft-setup-social-auth-164798203`
- Checkout `git checkout ft-setup-social-auth-164798203`
- Run `pipenv install` to install packages
- Run `python manage.py makemigrations`
- Run `python manage.py migrate`
- Run `python manage.py runserver`

   #### Testing Facebook
   - Make sure you are logged into your Facebook account. 
   - Visit [link](https://developers.facebook.com/) to create an app
   - Open this [link](https://developers.facebook.com/tools/explorer), generate a token and copy it

   #### Testing Google 
   - Make sure you are logged into your Google account
   - Open this link [link](https://developers.google.com/oauthplayground/)
   - In the 'scope' section select 'Google OAuth2 API v2' and check:white_check_mark: 
all the three fields in the dropdown
   - Click authorize the 'Authorize APIs' button and verify your google account
   - Click the 'Exchange authorization code for token' button
   - Copy the value of the 'access token' key
   
   #### Testing Twitter
   - Make sure you are logged into your Twitter account
   - Visit this [link](https://developer.twitter.com/en/apps) and create app
   -  Choose the 'keys and token' tab and copy both the access token


   #### Endpoints
   - Facebook - `POST 127.0.0.1:8000/api/users/login/facebook`
   - Google - `POST 127.0.0.1:8000/api/users/login/google`
   - Twitter - `POST 127.0.0.1:8000/api/users/login/google`


#### What are the relevant pivotal tracker stories?
[#164798203](https://www.pivotaltracker.com/story/show/164798203)


#### Screenshots
   #### Twitter
    Note: Twitter tokens are separated by a space

Twitter tokens
<img width="824" alt="Screenshot 2019-04-03 at 23 21 46" src="https://user-images.githubusercontent.com/35653907/55510897-f85d6000-5667-11e9-8466-4f9e6b09cf9e.png">
  
Post request
<img width="1055" alt="Screenshot 2019-04-03 at 23 54 32" src="https://user-images.githubusercontent.com/35653907/55512583-2c3a8480-566c-11e9-95ed-918e06d118af.png">

   #### Google
 Authorize APIs
<img width="752" alt="Screenshot 2019-04-03 at 23 17 19" src="https://user-images.githubusercontent.com/35653907/55510506-1d050800-5667-11e9-9f04-d1fabe9584a7.png">

Get Token
<img width="1395" alt="Screenshot 2019-04-03 at 23 59 04" src="https://user-images.githubusercontent.com/35653907/55512962-2e511300-566d-11e9-9a72-df1269c8c445.png">

Post request
<img width="1030" alt="Screenshot 2019-04-04 at 00 07 31" src="https://user-images.githubusercontent.com/35653907/55513215-da92f980-566d-11e9-9693-5362dd4c4d02.png">

   #### Facebook
Generate token
<img width="989" alt="Screenshot 2019-04-03 at 23 12 08" src="https://user-images.githubusercontent.com/35653907/55510132-39ed0b80-5666-11e9-8334-d915bf6c2a7a.png">

Post request
<img width="1039" alt="Screenshot 2019-04-04 at 00 10 22" src="https://user-images.githubusercontent.com/35653907/55513404-42494480-566e-11e9-9b19-0fc8fadbb6c8.png">


#### Postman Collection
[![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/bc400e4b5c7c7ae50050)
